### PR TITLE
Add assert_arrays_equal function

### PR DESCRIPTION
**Pull request by Pasifaee**
Added assert_arrays_equal function.

### CODE REVIEW

File: lib/src/lib.nr [hunk 1]
```diff
@@ -213,18 +195,44 @@ impl<PROOF_LEN, MAX_VALUE_LEN> TrieProof<20, PROOF_LEN, MAX_VALUE_LEN>
...
     }
 }
 
+/// Fills destination array with content of source array starting from the starting position.
+///
+/// # Arguments
+/// * `dest` - Destination array
+/// * `src` - Source array
+/// * `start_pos` - Starting position in source array 
+fn memcpy<N, M>(dest: &mut [u8; N], src: [u8; M], start_pos: u64)
```
Comment (LogvinovLeon) [inline lib/src/lib.nr]:
```suggestion
fn memcpy<N, M>(dest: &mut [u8; N], src: [u8; M], offset: u64)
```

File: lib/src/lib.nr [hunk 2]
```diff
@@ -213,18 +195,44 @@ impl<PROOF_LEN, MAX_VALUE_LEN> TrieProof<20, PROOF_LEN, MAX_VALUE_LEN>
...
+
+/// Verifies that specified number of elements from an array match elements from the offset array
+/// at specified offset.
+///
+/// # Arguments
+/// * `array` - Array to be verified
+/// * `offset_array` - Array to be verified against
+/// * `len` - Number of elements to be verified
+/// * `offset` - Offset in `offset_array`
+fn assert_arrays_equal<N, M>(array: [u8; N], offset_array: [u8; M], len: u64, offset: u64)
```
Comment (LogvinovLeon) [inline lib/src/lib.nr]:
```suggestion
fn assert_arrays_equal<N, M>(subarray: [u8; N], array: [u8; M], len: u64, offset: u64)
```

File: lib/src/lib.nr [hunk 3]
```diff
@@ -213,18 +195,44 @@ impl<PROOF_LEN, MAX_VALUE_LEN> TrieProof<20, PROOF_LEN, MAX_VALUE_LEN>
...
+{
+    for i in 0..N
+    {
+        if (i as u32) < (len as u32) {
```
Comment (LogvinovLeon) [inline lib/src/lib.nr]: Why do we need to cast to u32? Can't we compare on u64?

File: lib/src/lib.nr [hunk 4]
```diff
@@ -213,18 +195,44 @@ impl<PROOF_LEN, MAX_VALUE_LEN> TrieProof<20, PROOF_LEN, MAX_VALUE_LEN>
...
+            assert(array[i] == offset_array[offset + i]);
```
Comment (LogvinovLeon) [inline lib/src/lib.nr]: Assertion message

File: lib/src/lib.nr [hunk 5]
```diff
@@ -541,6 +549,31 @@ fn left_byte_shift<N>(input: [u8; N], n: u64) -> [u8; N]
...
+
+#[test]
+fn arrays_equal_test()
+{
+    let array = [3,4,5,6,7,8,9,10];
+    let offset_array = [1,2,3,4,5,6,7,8,9,10];
+    assert_arrays_equal(array, offset_array, 10, 2);
+}
+
+#[test(should_fail)]
```
Comment (LogvinovLeon) [inline lib/src/lib.nr]: With which message?

File: lib/src/lib.nr [hunk 6]
```diff
@@ -541,6 +549,31 @@ fn left_byte_shift<N>(input: [u8; N], n: u64) -> [u8; N]
     out
 }
 
+#[test]
+fn memcpy_test()
+{
+    let mut dest = [0; 10];
+    let src = [1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20];
```
Comment (LogvinovLeon) [inline lib/src/lib.nr]: Let's pick smaller arrays for tests. Arrays of size 2 and 3 are more than enough here and taking size 10 doesn't add to test robustness, but makes it harder to eyeball. Also - spaces after commas

File: lib/src/lib.nr [hunk 7]
```diff
@@ -541,6 +549,31 @@ fn left_byte_shift<N>(input: [u8; N], n: u64) -> [u8; N]
     out
 }
 
+#[test]
+fn memcpy_test()
```
Comment (LogvinovLeon) [inline lib/src/lib.nr]: Test names should start with test_. It's a convention. test_memcpy.

**Review by LogvinovLeon — CHANGES_REQUESTED** [review]

File: lib/src/lib.nr (hunk 3 above)
Comment (Pasifaee) [reply lib/src/lib.nr]: Ahmad casts to u32, not sure why though. It doesn't seem to be needed so I can uncast

File: lib/src/lib.nr (hunk 7 above)
Comment (Pasifaee) [reply lib/src/lib.nr]: Ahmad breaks this convention though

Comment (LogvinovLeon) [reply lib/src/lib.nr]: Ok

File: lib/src/lib.nr (hunk 6 above)
Comment (Pasifaee) [reply lib/src/lib.nr]: Ahmad doesn't use spaces after commas

**Pull request merged**